### ISSUE TITLE
Allow child component to alter what the parent component has rendered.

### ISSUE THREAD

**Issue by joshfoskett** [mjmlio/mjml#2107]
**Is your feature request related to a problem? Please describe.**

Essentially, I want to wrap the output of my custom component with an `if statement`. However, this is problematic because the parent component will wrap the child component with additional HTML. So if the` if statement` is `false`, the child component's HTML will not be rendered, but whatever the parent component wrapped around it will be. I understand that you can set the component as a `rawElement`, but this seemingly causes more issues than it solves in this context.

So to give you an example, I've created a custom component called `MjDynamicImage`.
```
class MjDynamicImage extends BodyComponent {
  render () {
    const renderedHtml = this.renderMJML('<mj-image src="{{ image_url }}"></mj-image>')

    return `
      {% if image_url %}
        ${renderedHtml}
      {% endif %}
    `
  }
}
```

If `mj-dynamic-image` is wrapped with `mj-column`, and `image_url` is `falsey`, this still gets rendered:
```
            <tr>
              <td
                 align="center" style="font-size:0px;padding:10px 25px;word-break:break-word;"
              >
              </td>
            </tr>
```
https://github.com/mjmlio/mjml/blob/master/packages/mjml-column/src/index.js#L254

**Describe the solution you'd like**

The ability for the child component to access/alter what the parent component rendered for it.

This could be accomplished by doing something like this in [`packages/mjml-core/src/createComponent.js#L228`](https://github.com/mjmlio/mjml/blob/master/packages/mjml-core/src/createComponent.js#L228).
```
        if (component !== null) {
          let content = renderer(component);

          if (typeof component.parentRender === 'function') {
            content = component.parentRender(content);
          }

          output += content;
        }
```
Which would allow me to do something like:
```
class MjDynamicImage extends BodyComponent {
  parentRender(content) {
    return `
      {% if image_url %}
        ${content}
      {% endif %}
    `
  }

  render () {
    return this.renderMJML('<mj-image src="{{ image_url }}"></mj-image>')
  }
}
```

**Describe alternatives you've considered**

An alternative way to accomplish this would be to simply alter the MJML before passing it to `mjml2html`. By that I mean, I could convert the MJML to JSON, find all instances of `mj-dynamic-image`, and replace it with `mj-image`, and then place `mj-raw` above and below it.

So for instance, this:
```
<mj-column>
  <mj-dynamic-image>
  <mj-text>Some text.</mj-text>
</mj-column>
```

Would become:
```
<mj-column>
  <mj-raw>
    {% if image_url %}
  </mj-raw>
  <mj-image src="{{ image_url }}" />
  <mj-raw>
    {% endif %}
  </mj-raw>
  <mj-text>Some text.</mj-text>
</mj-column>
```

However, this feels a bit hacky in comparison to creating an actual component.

Another option as mentioned above would be to register a custom component and set `rawElement` to `true`. This would make it so that the parent component doesn't wrap anything specifically around the child component. However, you'd then need context as to what the parent component is in the child component, and you'd essentially have recreate what the parent component was doing in terms of wrapping it, which has the potential to break over time if there are any changes, as they wouldn't be reflected.

**Additional context**

TL;DR: I have a custom component, which I want to wrap an `if statement` around, but the parent component will wrap additional HTML around the `if statement`, which results in part of the component still being rendered when the `if statement ` is `falsey` ([resulting in empty space in the template](https://d.joshf.xyz/ZYOkuT)).

**Comment by iRyusa** [comment]
You could do something similar by just making dynamic image as a raw component here, and provide the full tr + td + surrounded by if/else templating. Only "downside" is you have to bind by yourself the padding to tr/td but that's not really a big thing as your solution would need that too.

I'm not really a big fan of adding parentRenderer because how limited the use case would be here and that can be done today with raw component

**Comment by joshfoskett** [comment]
@iRyusa I appreciate the quick reply.

Is it possible for a child component to know what the parent component is on `render`? I assumed that it may be provided in `context`, but that doesn't appear to be the case.

```
{ addMediaQuery: [Function: addMediaQuery],
  addHeadStyle: [Function: addHeadStyle],
  addComponentHeadSyle: [Function: addComponentHeadSyle],
  setBackgroundColor: [Function: setBackgroundColor],
  processing: [Function: processing],
  containerWidth: '300px' }
```

I ask this because it appears that `mj-image` can be used in both `mj-column` and `mj-hero`, which means that I'd have to conditionally wrap the component based on the parent.

`mj-column`:

```jsx
            <tr>
              <td
                ${component.htmlAttributes({
                  align: component.getAttribute('align'),
                  'vertical-align': component.getAttribute('vertical-align'),
                  class: component.getAttribute('css-class'),
                  style: {
                    background: component.getAttribute(
                      'container-background-color',
                    ),
                    'font-size': '0px',
                    padding: component.getAttribute('padding'),
                    'padding-top': component.getAttribute('padding-top'),
                    'padding-right': component.getAttribute('padding-right'),
                    'padding-bottom': component.getAttribute('padding-bottom'),
                    'padding-left': component.getAttribute('padding-left'),
                    'word-break': 'break-word',
                  },
                })}
              >
                ${component.render()}
              </td>
            </tr>
```

`mj-hero`:

```jsx
                    <tr>
                      <td
                        ${component.htmlAttributes({
                          align: component.getAttribute('align'),
                          background: component.getAttribute(
                            'container-background-color',
                          ),
                          class: component.getAttribute('css-class'),
                          style: {
                            background: component.getAttribute(
                              'container-background-color',
                            ),
                            'font-size': '0px',
                            padding: component.getAttribute('padding'),
                            'padding-top': component.getAttribute(
                              'padding-top',
                            ),
                            'padding-right': component.getAttribute(
                              'padding-right',
                            ),
                            'padding-bottom': component.getAttribute(
                              'padding-bottom',
                            ),
                            'padding-left': component.getAttribute(
                              'padding-left',
                            ),
                            'word-break': 'break-word',
                          },
                        })}
                      >
                        ${component.render()}
                      </td>
                    </tr>
```

I'd seemingly have to conditionally render those (if I took the `rawElement` approach), which is fine, as I imagine it's rare for those to change.

The advantage of something like `parentRender` is that if they did change, it'd automatically be reflected with no changes on my end. But I can see that it does have the potential to open up a can of worms, for something as niche as this particular use case.

**Comment by iRyusa** [comment]
I think this is a bit hard to do because you doesn't know what the parent could be. The best way to handle that is to use the rawElement as you would have total control of it !

I'm closing for now but if this became a highly needed feature let's open up a discussion about the solution you provide here 👍
